src/main/proto/netflix/titus/titus_cluster_membership_api.proto

```diff
@@ -0,0 +1,205 @@
+    map<string, string> labels = 7;
+}
+
+message ClusterMembershipEvent {
+    message ServiceInstanceUpdate {
+        ChangeRecord change = 1;
+        ServiceInstance previous = 2;
+        ServiceInstance after = 3;
+    }
+
+    message ServiceInstanceRemoved {
+        ChangeRecord change = 1;
+        ServiceInstance last = 2;
+    }
+
+    oneof Event {
+        ServiceInstanceUpdate serviceUpdated = 1;
+        ServiceInstanceRemoved serviceRemoved = 2;
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GRPC services
+
+message Id {
+    string id = 1;
+}
+
+message ServiceInstances {
+    repeated ServiceInstance instances = 1;
+}
+
+message ChangeLogRequest {
+    // Filtering criteria:
+    // * 'ids' - comma separated list of instance ids
+    // * 'instanceLimit' - maximum number of records to return per instance
+    // * 'limit' - maximum number of records to return
+    // * 'before' - events before the given timestamp (epoch)
+    // * 'since' - events at or after the given timestamp (epoch)
+    map<string, string> filters = 1;
+
+    string fields = 2;
+}
+
+message ChangeLogResponse {
+    message ChangeRecord {
+        ChangeRecord change = 1;
+        ServiceInstance changedTo = 2;
+    }
+
+    repeated ChangeRecord changes = 1;
+}
+
+message UpdateServiceLabelsRequest {
+    string instanceId = 1;
+
+    map<string, string> labels = 2;
+}
+
+message DeleteServiceLabelsRequest {
+    string instanceId = 1;
+    repeated string keys = 2;
+}
+
+message UpdateServiceStateRequest {
+    string instanceId = 1;
+
+    string reasonMessage = 2;
+
+    ServiceInstance.ServiceState newState = 3;
+}
+
+service ClusterMembershipService {
+
+    // Get all known service instances.
+    rpc GetServiceInstances (google.protobuf.Empty) returns (ServiceInstances) {
+    }
+
+    // Get service instance with the given id.
+    rpc GetServiceInstance (Id) returns (ServiceInstance) {
+    }
+
+    // Get recent changes.
+    rpc GetChangeLog (ChangeLogRequest) returns (ChangeLogResponse) {
+    }
+
+    // Adds all labels from the request object to the target service instance. Labels that exist are
+    // overridden. Returns the updated object.
+    rpc UpdateServiceLabels (UpdateServiceLabelsRequest) returns (ServiceInstance) {
+    }
+
+    // Removes all specified labels from the target object. Labels that do not exist are ignored.
+    // Returns the updated object.
+    rpc DeleteServiceLabels (DeleteServiceLabelsRequest) returns (ServiceInstance) {
+    }
+
+    // Change state of a given service instance. Returns the updated object.
+    rpc UpdateServiceState (UpdateServiceStateRequest) returns (ServiceInstance) {
+    }
+
+    // Requests the service instance that handles this request to stop being leader. If the given instance
+    //  is not a leader, the request is ignored.
+    rpc StopBeingLeader (google.protobuf.Empty) returns (google.protobuf.Empty) {
```

Comment:
Do we need a becomeLeader() as well. Based on the implementation it can throw an exception for invalid state transition, violating invariants etc.

Reply:
No, leader election process is handled internally. Choosing a specific instance to be a leader can be achieved by inactivating all other instances, and making the call `StopBeingLeader` to the current leader. It is not something we want to do in the production system, but may be useful for testing.

Reply:
do we control which instances are eligible for leadership through labels only? I wonder if we need a dedicated API for it.

Reply:
We control this explicitly by managing member state: ```enum State {

        // Member instance is starting, and its operational state is not known yet.
        Starting = 0;

        // Member is inactive, and does not participate in the leader election process. Predefined
        // reason codes in this state:
        // * 'outOfService' - the member is healthy, but taken of out of service by an administrator
        // * 'unhealthy' - the member is not healthy
        Inactive = 1;

        // Member is fully activated, and if does not participate in the leader election process, is ready to
        // take traffic.
        Active = 2;
    }
```

and `UpdateMemberState` method.